### PR TITLE
Allow configuring endpoint for Azure FS

### LINKED ISSUE #21307: Allow Azure govcloud for native Azure file system
https://trinodb.slack.com/archives/C0305TQ05KL/p1711623700672519

Azure gov cloud uses https://%s.blob.core.usgovcloudapi.net  not https://%s.blob.core.windows.net , 

The validation does not allow this and needs to be changed to be more flexible.

https://github.com/trinodb/trino/blob/9ec2b07eddb88838eeb243e54df16d323ec4a0fb/lib/trino-filesystem-azure/src/main/java/io/trino/filesystem/azure/AzureLocation.java#L26

private static final String INVALID_LOCATION_MESSAGE = "Invalid Azure location. Expected form is 'abfs://[<containerName>@]<accountName>.dfs.core.windows.net/<filePath>': %s";

### PR DESCRIPTION
**Pull request by electrum**
## Additional context and related issues

Fixes #21307

## Release notes

(x) Release notes are required, with the following suggested text:

```markdown
# Delta Lake
* Allow configuring endpoint for the native Azure filesystem. ({issue}`23071`)

# Hive
* Allow configuring endpoint for the native Azure filesystem. ({issue}`23071`)

# Hudi
* Allow configuring endpoint for the native Azure filesystem. ({issue}`23071`)

# Iceberg
* Allow configuring endpoint for the native Azure filesystem. ({issue}`23071`)
```

### CODE REVIEW

File: lib/trino-filesystem-azure/src/main/java/io/trino/filesystem/azure/AzureFileSystemConfig.java [hunk 1]
```diff
@@ -47,6 +48,19 @@ public AzureFileSystemConfig setAuthType(AuthType authType)
         return this;
     }
 
+    @NotEmpty
+    public String getEndpoint()
+    {
+        return endpoint;
+    }
+
+    @Config("azure.endpoint")
```
Comment (mosabua) [inline lib/trino-filesystem-azure/src/main/java/io/trino/filesystem/azure/AzureFileSystemConfig.java:57]: We should document this, including the default value and alternative example value for govcloud.

File: docs/src/main/sphinx/object-storage/file-system-azure.md [hunk 2]
```diff
@@ -27,6 +27,12 @@ system support:
     authentication used with `NONE`. Use `ACCESS_KEY` for
     [](azure-access-key-authentication) or and `OAUTH` for
     [](azure-oauth-authentication).
+* - `azure.endpoint`
+  - Hostname suffix of the Azure storage endpoint.
+    Defaults to `core.windows.net` for the global Azure cloud.
+    Use `core.usgovcloudapi.net` for the Azure US Government cloud,
+    `core.cloudapi.de` for the Azure Germany cloud,
+    or `core.chinacloudapi.cn` for the Azure China cloud.
```
Comment (mosabua) [inline docs/src/main/sphinx/object-storage/file-system-azure.md:35]: Is there a link to a list in case they add more?

**Review by mosabua — APPROVED** [review]
Looks good to me now.

File: docs/src/main/sphinx/object-storage/file-system-azure.md (hunk 2 above)
Comment (mosabua) [reply docs/src/main/sphinx/object-storage/file-system-azure.md:35]: I could not find any and this list is good enough to start.

File: lib/trino-filesystem-azure/src/main/java/io/trino/filesystem/azure/AzureLocation.java [hunk 3]
```diff
@@ -32,7 +32,16 @@ class AzureLocation
...
     private final String account;
+    private final String endpoint;
 
+    /**
+     * Creates a new location based on the endpoint, storage account, container and blob path parsed from the location.
+     * <p>
+     * Locations follow the conventions used by
+     * <a href="https://docs.microsoft.com/en-us/azure/storage/blobs/data-lake-storage-introduction-abfs-uri">ABFS URI</a>
+     * that follows the following convention
+     * <pre>{@code abfs://<container-name>@<storage-account-name>.dfs.<endpoint>/<blob_path>}</pre>
```
Comment (findinpath) [inline lib/trino-filesystem-azure/src/main/java/io/trino/filesystem/azure/AzureLocation.java:43]:
```suggestion
     * Creates a new location based on the endpoint, storage account, container and blob path parsed from the location.
     * <p>
     * Locations follow the conventions used by
     * <a href="https://docs.microsoft.com/en-us/azure/storage/blobs/data-lake-storage-introduction-abfs-uri">ABFS URI</a> syntax:
     * <pre>{@code abfs://<container-name>@<storage-account-name>.dfs.<endpoint>/<blob_path>}</pre>
```

Comment (findinpath) [reply lib/trino-filesystem-azure/src/main/java/io/trino/filesystem/azure/AzureLocation.java:43]: `Locations follow the conventions ... that follows the following convention`

Too many follow

**Review by findinpath — APPROVED** [review]

**Review by ms1111 — COMMENTED** [review]
What would you think about making the entire endpoint URL configurable, so instead of hardcoding `<account name>` + `.dfs.` + `<endpoint>`, a user could configure `https://<accountname>.dfs.<endpoint>` directly?

That would open things up to test with use other implementations like the Azurite emulator in the future, where you're likely to need something like `http://<ip or container name or localhost>:some port`, with no `.dfs.`

File: lib/trino-filesystem-azure/src/main/java/io/trino/filesystem/azure/AzureFileSystemConfig.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (electrum) [reply lib/trino-filesystem-azure/src/main/java/io/trino/filesystem/azure/AzureFileSystemConfig.java:57]: That's a good suggestion, but I think `endpoint` is fine, as this is quite specific and will require reading the documentation to understand it.

**Comment by electrum** [comment]
@ms1111 That's a good idea, which I considered, but it's complicated due to the [ABFS URI syntax](https://learn.microsoft.com/en-us/azure/storage/blobs/data-lake-storage-introduction-abfs-uri) containing the storage account. @dain tried using Azurite when he wrote the Azure FS, but ran into various issues and incompatibilities with the real Azure storage, which is why we don't use it for testing.

We can extend the config in the future when there is a real use case or viable alternate implementation. Let us know if you have a specific use case in mind. The goal of the current change is to support all of the Azure clouds while being easy for users to configure.

**Comment by ms1111** [comment]
@electrum Yeah, I was following some of the Azurite issues in https://github.com/Azure/Azurite/issues/553 and I figure if they're solved in the future, then it would require changing the Trino configuration again to allow specifying a local hostname, port, and scheme (http).

I was just thinking that might be a hard sell later, because you'd end up with two different historical ways to override the endpoint - the one being added here, and a new way to allow overriding the other pieces.

Being able to use Azurite is handy for local & CI testing if you have a software project that uses Trino and other components that all need to talk to a data lake. You can write integration tests that hit a few different containers and they all can touch the same tables. (Sometimes you can do this in with file:// URIs instead of emulating abfs:// or s3a://, but not always - there's always some weird limitation, like for example CircleCI's remote docker executor can't do bind mounts, so you can't share files between the primary and secondary containers.)

Because of the hardcoded endpoint URL in Trino, we're trying with S3/MinIO instead, but given we're an Azure shop, it'd be nice to be able to do it all with the Azure filesystem eventually.

**Comment by electrum** [comment]
@ms1111 We can extend the config in the future to accept a URL. This won't conflict with the existing configuration.

However, I don't think that Azurite integration would be as simple as allowing to configure a URL endpoint, as the way [storage accounts](https://github.com/Azure/Azurite?tab=readme-ov-file#storage-accounts
) are handled when constructing URLs is different. We need to take the storage account from the ABFS location and generate an endpoint URL, which has a different format for real Azure vs Azurite. So if Azurite ever actually supports ADLS, we will likely need to add specific support for it.

**Comment by electrum** [comment]
We test the S3 file system against MinIO, S3Mock and LocalStack. I agree the local storage tests are very useful and we will certainly support Azurite if they ever implement ADLS support and make it actually compatible with Azure.

**Comment by mosabua** [comment]
Lets ship this @electrum

**Comment by mosabua** [comment]
I am merging after chatting with @anusudarsan

**Pull request merged**